src/Env/EnvFactory.php

```diff
@@ -27,8 +29,8 @@ class EnvFactory
      *
      * @return Env
      */
-    public static function createEnv(EnvEnum $type)
+    public static function createEnv(EnvEnum $type, AppVendor $appVendor, Dependency ...$dependencies)
```

Comment:
> i.e. do not have to iterate over all elements of an array for being sure they are of the good type

Sure. But that's a limitation of PHP to not be able to valid what is in an array (generics may help regarding this in PHP 7.2 or later). 
But variadics aim is not about validation, and thus are misused if you only use variadics for this reason.